examples/vision/patch_conv_net.py

```diff
@@ -0,0 +1,661 @@
+            }
+        )
+        return config
+
+    def build(self, input_shape):
+        self.attention = layers.MultiHeadAttention(
+            num_heads=1, key_dim=self.dimensions, dropout=0.2,
+        )
+        self.layer_norm1 = layers.LayerNormalization(epsilon=1e-6)
+        self.layer_norm2 = layers.LayerNormalization(epsilon=1e-6)
+        self.layer_norm3 = layers.LayerNormalization(epsilon=1e-6)
+        self.mlp = keras.Sequential(
+            [
+                layers.Dense(units=self.dimensions, activation=tf.nn.gelu),
+                layers.Dropout(0.2),
+                layers.Dense(units=self.dimensions, activation=tf.nn.gelu),
+            ]
+        )
+        self.dense = layers.Dense(units=self.num_classes)
+        self.flatten = layers.Flatten()
+
+    def call(self, x):
+        batch_size = tf.shape(x)[0]
+        # Expand the class token batch number of times.
+        class_token = tf.repeat(self.cls, repeats=batch_size, axis=0)
+        # Concat the input with the trainable class token.
+        x = tf.concat([class_token, x], axis=1)
+        # Apply attention to x.
+        x = self.layer_norm1(x)
+        x, viz_weights = self.attention(
+            query=x[:, 0:1], key=x, value=x, return_attention_scores=True
+        )
+        class_token = class_token + x
+        class_token = self.layer_norm2(class_token)
+        class_token = self.flatten(class_token)
+        class_token = self.layer_norm3(class_token)
+        class_token = class_token + self.mlp(class_token)
+        # Build the logits
+        logits = self.dense(class_token)
+        return logits, tf.squeeze(viz_weights)[..., 1:]
+
+
+"""
+## Patch Conv Net
+
+The patch-conv-net is shown in the figure below.
+
+| ![image model](https://i.imgur.com/NHiQeac.png) |
+| :--: |
+| [Source](https://arxiv.org/abs/2112.13692) |
+
+All the modules in the architecture are built in the earlier seciton.
+In this section, we will stack all of the different modules together.
+"""
+
+
+class PatchConvNet(keras.Model):
+    def __init__(
+        self,
+        stem,
+        trunk,
+        attention_pooling,
+        train_augmentation_model,
+        test_augmentation_model,
+        **kwargs,
+    ):
+        super().__init__(**kwargs)
+        self.stem = stem
+        self.trunk = trunk
+        self.attention_pooling = attention_pooling
+        self.train_augmentation_model = train_augmentation_model
+        self.test_augmentation_model = test_augmentation_model
+
+    def get_config(self):
+        config = super().get_config()
+        config.update(
+            {
+                "stem": self.stem,
+                "trunk": self.trunk,
+                "attention_pooling": self.attention_pooling,
+                "train_augmentation_model": self.train_augmentation_model,
+                "test_augmentation_model": self.test_augmentation_model,
+            }
+        )
+        return config
+
+    def calculate_loss(self, inputs, test=False):
```

Comment:
I would recommend making this method private, because it risks confusion with `compute_loss()` which is an overridable method.